### DISCUSSION TITLE
Adding Interactive Server, WebAssembly, Auto Modes to Oqtane Framework

### DISCUSSION THREAD

**Discussion by thabaum** [oqtane/oqtane.framework#3407] [General]
I am attempting working with the custom router in Oqtane to implement the changes needed to allow Interactive Auto Mode or Interactive Server Mode or Interactive WebAssembly Mode within the Oqtane Framework.

For dealing with the new feature this is where things do not get accepted by the build process after adding the other parts and using the NET8 branch:

In the Server project Startup.cs I added the following:

```
            services.AddRazorComponents()
                .AddInteractiveServerComponents()
                .AddInteractiveWebAssemblyComponents();

```

Which now is recognized thanks to the package references using latest preview which I had issue with earlier.

Now just left with the following error:

`'IApplicationBuilder' does not contain a definition for 'MapRazorComponents' and the best extension method overload 'RazorComponentsEndpointRouteBuilderExtensions.MapRazorComponents<App>(IEndpointRouteBuilder)' requires a receiver of type 'Microsoft.AspNetCore.Routing.IEndpointRouteBuilder'`

When adding these lines to the Startup.cs

```
            app.MapRazorComponents<App>()
                .AddInteractiveServerRenderMode()
                .AddInteractiveWebAssemblyRenderMode();
```

[Microsoft.AspNetCore.Routing.IEndpointRouteBuilder](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.routing.iendpointroutebuilder?view=aspnetcore-8.0)

What logic replaces or relates to this `receiver` needed in Oqtane?

Is this going to be in the `App.razor` needing a parameter and logic for `IEndpointRouteBuilder` in order to handle this feature?

Just curious what we can do to make this feature work, or at least try breaking things using it for self-interest.  I am trying to understand how to resolve this error properly if possible. Thought I would get a discussion going in the event I can help with any insights to figure this thing out prior to .NET 8 release.

**Comment by sbwalker** [comment]
@thabaum it is going to take some significant research to understand how the new render modes truly work under the covers in .NET 8. I am fairly sure there is no documentation available - the only documentation for .NET 8 is high level marketing information on how to build a Hello World app using the new Blazor Web template. So this will require trial and error experimentation and reviewing the Blazor source code on GitHub (thank goodness it is open source). Technically I don't even know if it is possible - it depends on whether certain scenarios were even considered during development of the new Blazor features in .NET 8 (keeping in mind that Oqtane's dynamic modular approach is completely different than the standard Blazor development approach). And conceptually I am still trying to wrap my head around some practical aspects of these new features. For example, since most themes have a Login button which is "interactive" by nature (as it needs to support an onclick event) then does it mean that all page renderings for a site need to be "interactive" as well - which would basically mean (static) Server-Side Rendering is not useful in Oqtane (it may be easier to test a scenario like this using the basic Blazor Web template rather than from within Oqtane).

**Comment by sbwalker** [comment]
@thabaum to more specifically answer your questions above, it is my current understanding that in .NET 8 the Blazor engine scans assemblies during startup to identify components. It builds a route table based on these components. It then passes the route table from the server to the client so that it can be used in the default Blazor router. The router compares the current route to the route table and makes a determination if it should short-circuit processing (non-interactive) or continue processing (interactive). What I am not sure of is how it identifies components - originally I thought it looked for the @page directive and captured the route value. This may be correct however there would still need to be some other attribute to determine if a component is intended to be interactive or not.

Regardless, Oqtane does not support the @page directive in components as routes are not static - they are dynamic and are stored in the database. Oqtane does not use the default Blazor router - it uses a custom router. So there are a LOT of major architectural differences which need to be sorted out.

**Reply by thabaum** [reply]
@sbwalker thanks for the insights.  I do recall the Oqtane router being custom and NOT using the `@page` directive as you pointed out.  This is why I am trying to understand both architectures (Oqtane and .NET Blazor) for the way they handle the routing tables along with the circuit handling when requests are made.  I am wondering if you can point out all the files in Oqtane that relate to this architecture so I can get a complete understanding by reviewing the code in detail that relates.  I want to go through Oqtane's and .NET Blazors differences to see how they can meet in the middle if possible and help come up with a solution to enhance Oqtane using what they have provided in the Blazor API's.  I did notice the login and few other components or pages have this `@page`directive in the Oqtane Framework.

I am going to make a branch to explicitly work on this and I sleep on it while thinking out possibilities.  It is nothing major to have in Oqtane, but a challenge and should help me understand both frameworks more in depth out of my own personal curiosities.

The `Oqtane.Server.Pages` space has pages that use the `@page` directive and I have been reviewing `Oqtane.Server.Infrastructure.Interfaces` areas for getting an understanding.  If you can give me a list of files that relate to the Oqtane router so I can be sure to review these areas while testing out ideas that would be helpful.  I just want to be sure I am looking at all relating files for potential possibilities.

If I understand correctly the `@page` directive is ignored on these pages in the "Pages" folder.  I did want to ask to be sure however.  This was the area I thought Oqtane could potentially add or test some of these interactive mode capabilities.

I would love to see "Auto" mode be enabled/disabled for the entire application was a first goal, then per component or dynamic page.

I am visioning a setting for each site/page that can be set programmatically within the Oqtane Framework to set the interactive modes.  But what I really would like to harness is the server initial connection with downloading the WebAssembly dll in background if connection and browser supports this for the entire app.  The component/page interactive modes I am less concerned about at this time. To accomplish this of course I would need a very in-depth understanding of both Oqtane and Blazor router logic.  I may not get the solution, but I may save some troubles sharing insights on what doesn't work.

This [https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0) has been my resource for reviewing how things work and relate.

And the section link below is what I have been testing setting up in Oqtane:

[https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#enable-support-for-interactive-render-modes](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#enable-support-for-interactive-render-modes)

Issue is:

```
app.MapRazorComponents<App>()
    .AddInteractiveServerRenderMode()
    .AddInteractiveWebAssemblyRenderMode();
```

Has an error as `IApplicationBuilder` does have what is needed to work and the overload requires a receiver:

`Error	CS1929	'IApplicationBuilder' does not contain a definition for 'MapRazorComponents' and the best extension method overload 'RazorComponentsEndpointRouteBuilderExtensions.MapRazorComponents<App>(IEndpointRouteBuilder)' requires a receiver of type 'Microsoft.AspNetCore.Routing.IEndpointRouteBuilder'	Oqtane.Server	C:\...\NET8\oqtane.framework-net8\Oqtane.Server\Startup.cs	202	Active
`
So I am trying to investigate this area first.

Thinking out loud a little here maybe Oqtane needs a custom IEndpointRouteBuilder that is a custom copy the Blazor router version or maybe through dependency injection add it's capabilities?

For a per component like the login button not sure if I am thinking right here but adding the intereactive mode attribute is described more here:

[https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#apply-a-render-mode-to-a-component-definition](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#apply-a-render-mode-to-a-component-definition)

For example `@attribute [RenderModeInteractiveAuto]` can be added to a component.

But there are precautions not to use this on everything everywhere (similar to rendering 1000's of button components)... only in certain components should these attributes be added and not all within the application actually will use this render mode feature properly or you can run into issues.

This section also explains more on the Auto feature

[https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#auto-render-mode](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#auto-render-mode)

I like the auto mode as it should dictate the proper use for components rendering and avoid errors in the process.  Different components have different acceptable uses and letting Blazor decide this seems to be the smoothest choice, but maybe not the best if it takes more time to render as the framework needs to make more choices along the way.

Since this is all new, I don't expect an answer from anyone, not even Microsoft as I am not sure they know 100% what bugs are coming from this along with security issues if any.  It is more of a pioneering of these concepts available in Blazor .NET 8 to introduce into Oqtane as they mature in the upcoming month(s) ahead.

The other thing I need to figure out is how to properly test these which I need to watch the video of Dan Roth explaining and showing in detail how to recognize these features are in fact working.  I can trust they work in Blazor applications, but to see them working in Oqtane's Blazor application I was thinking Auto Mode would give me the best view as you go on a page first that is server mode, hit refresh a little laters and you should see a WebAssembly dll from my understanding in developer tools.

Here is a link to his blog that relates to the latest release: 

[https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-rc-2/](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-rc-2/)

And this page gives us some insights to the new and obsolete features to prepare for which can also give clues while investigating:

[https://learn.microsoft.com/en-us/dotnet/core/compatibility/8.0#aspnet-core](https://learn.microsoft.com/en-us/dotnet/core/compatibility/8.0#aspnet-core)

I just watched the VS RC 2 release on YouTube as well [https://www.youtube.com/watch?v=C2y9tXcbweg](https://www.youtube.com/watch?v=C2y9tXcbweg)

This gave a lot of insight to where things are at, but I feel things are still going to change some in coming weeks after watching.  The `@attribute` wont be used instead `@rendermode` will be used instead, so documentation needs updated I believe here as it points to using the @attribute

And as you said I think I need to build a few `Blazor Web App` projects and play with these features here more reviewing all the startup logic needed.

There is another feature for forms as well for "data enhanced" attributes on forms like settings to keep data from being removed from a page when submitting data.  This needs further review as well for modules and components within the Oqtane Framework.

Also another note I thought might apply here to Oqtane is component libraries are already implementing these render modes in their UI Libraries.
